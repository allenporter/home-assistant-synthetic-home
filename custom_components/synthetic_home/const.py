"""Constants for Synthetic Home."""

DOMAIN = "synthetic_home"
DEFAULT_NAME = "Synthetic Home"
CONF_FILENAME = "config_filename"

# Used for loading initial state of an entity and it persists across config
# entry reloads. Updated with service calls.
DATA_DEVICE_STATES = "device_states"

ATTR_AREA_NAME = "area"
ATTR_DEVICE_NAME = "device"
ATTR_DEVICE_STATE_KEY = "device_state_key"
ATTR_CONFIG_ENTRY_ID = "config_entry_id"
