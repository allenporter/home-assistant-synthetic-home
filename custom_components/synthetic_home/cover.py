"""Cover platform for Synthetic Home."""

import datetime
from typing import Any

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, callback, CALLBACK_TYPE
from homeassistant.components.cover import (
    CoverEntity,
    CoverDeviceClass,
    CoverEntityFeature,
    ATTR_POSITION,
    DOMAIN as COVER_DOMAIN,
)
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.event import async_track_time_interval

from .const import DOMAIN
from .entity import SyntheticDeviceEntity
from .model import ParsedDevice

COVER_STEP = 10
COVER_STEP_TIME = datetime.timedelta(seconds=1)


async def async_setup_entry(
    hass: HomeAssistant, entry: ConfigEntry, async_add_devices: AddEntitiesCallback
) -> None:
    """Set up cover platform."""
    synthetic_home = hass.data[DOMAIN][entry.entry_id]

    async_add_devices(
        SyntheticCover(device, entity.entity_key, **entity.attributes)
        for device in synthetic_home.devices
        for entity in device.entities
        if entity.platform == COVER_DOMAIN
    )


class SyntheticCover(SyntheticDeviceEntity, CoverEntity):
    """synthetic_home cover class."""

    # These are the attributes that represent the cover state
    _attr_is_closing = False
    _attr_is_closed = True
    _attr_is_opening = False
    # The position is used to infer the other attributes (0 is closed, 100 is open)
    _attr_current_cover_position: int = 0

    _target_cover_position: int | None = None
    _timer_unsub: CALLBACK_TYPE | None = None

    def __init__(
        self,
        device: ParsedDevice,
        key: str,
        supported_features: CoverEntityFeature,
        *,
        device_class: CoverDeviceClass | None = None,
        state: bool | None = None,
    ) -> None:
        """Initialize the SyntheticCover."""
        super().__init__(device, key)
        self._attr_supported_features = supported_features
        if device_class:
            self._attr_device_class = device_class
        if state:
            self._attr_current_cover_position = 100

    async def async_will_remove_from_hass(self) -> None:
        """When entity will be removed from Home Assistant."""
        if self._timer_unsub is not None:
            self._timer_unsub()
            self._timer_unsub = None

    async def async_close_cover(self, **kwargs: Any) -> None:
        """Close the cover."""
        if self._attr_current_cover_position == 0:
            # Already closed
            return
        self._attr_is_closing = True
        self._attr_is_opening = False
        self._target_cover_position = 0
        self._start_moving()
        self.async_write_ha_state()

    async def async_open_cover(self, **kwargs: Any) -> None:
        """Open the cover."""

        if self._attr_current_cover_position == 1000:
            # Already open
            return

        # The following attributes are set
        self._attr_is_closing = False
        self._attr_is_opening = True
        self._target_cover_position = 100
        self._start_moving()
        self.async_write_ha_state()

    async def async_set_cover_position(self, **kwargs: Any) -> None:
        """Move the cover to a specific position."""
        position: int = kwargs[ATTR_POSITION]
        if self._attr_current_cover_position == position:
            # Nothing to do
            return
        self._target_cover_position = position
        self._attr_is_closing = (
            self._target_cover_position < self._attr_current_cover_position
        )
        self._attr_is_opening = (
            self._target_cover_position > self._attr_current_cover_position
        )
        self._start_moving()
        self.async_write_ha_state()

    async def async_stop_cover(self, **kwargs: Any) -> None:
        """Stop the cover."""
        self._is_closing = False
        self._is_opening = False
        if self._timer_unsub is not None:
            self._timer_unsub()
            self._timer_unsub = None
            self._target_cover_position = None

    @callback
    def _start_moving(self) -> None:
        """Start moving the cover."""
        if self._timer_unsub is None:
            self._timer_unsub = async_track_time_interval(
                self.hass,
                action=self._move_cover,
                interval=COVER_STEP_TIME,
            )

    async def _move_cover(self, now: datetime.datetime) -> None:
        """Track time changes."""
        if (
            self._target_cover_position is not None
            and self._attr_current_cover_position > self._target_cover_position
        ):
            self._attr_current_cover_position -= COVER_STEP
            self._attr_current_cover_position = max(
                self._attr_current_cover_position, self._target_cover_position
            )
        elif (
            self._target_cover_position is not None
            and self._attr_current_cover_position < self._target_cover_position
        ):
            self._attr_current_cover_position += COVER_STEP
            self._attr_current_cover_position = min(
                self._attr_current_cover_position, self._target_cover_position
            )
        else:
            # Reached target
            self._attr_is_closing = False
            self._attr_is_opening = False
            self._attr_is_closed = self._attr_current_cover_position == 0
            await self.async_stop_cover()
        self.async_write_ha_state()
