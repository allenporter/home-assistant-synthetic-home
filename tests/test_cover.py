"""Test Synthetic Home cover."""

import datetime

import pytest
from syrupy import SnapshotAssertion

from homeassistant.const import Platform
from homeassistant.components.cover import (
    DOMAIN as COVER_DOMAIN,
    SERVICE_OPEN_COVER,
    SERVICE_CLOSE_COVER,
    SERVICE_SET_COVER_POSITION,
    ATTR_POSITION,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant
from homeassistant.util import dt as dt_util
from pytest_homeassistant_custom_component.common import async_fire_time_changed

from pytest_homeassistant_custom_component.common import MockConfigEntry

from .conftest import FIXTURES, restore_state


@pytest.fixture(name="platforms")
def mock_platforms() -> list[Platform]:
    """Set up platform."""
    return [Platform.COVER]


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/smart-blinds-example.yaml", "cover.left_shade")],
)
async def test_smart_blinds(
    hass: HomeAssistant, setup_integration: None, test_entity: str
) -> None:
    """Test smart blinds which are a positionable cover."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closed"
    assert state.attributes == {
        "friendly_name": "Left Shade",
        "device_class": "blind",
        "current_position": 0,
        "supported_features": 7,
    }

    await hass.services.async_call(
        COVER_DOMAIN,
        SERVICE_SET_COVER_POSITION,
        service_data={ATTR_ENTITY_ID: test_entity, ATTR_POSITION: 54},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "opening"

    for _ in range(7):
        future = dt_util.utcnow() + datetime.timedelta(seconds=1)
        async_fire_time_changed(hass, future)
        await hass.async_block_till_done()

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "open"
    assert state.attributes == {
        "friendly_name": "Left Shade",
        "device_class": "blind",
        "current_position": 54,
        "supported_features": 7,
    }

    await hass.services.async_call(
        COVER_DOMAIN,
        SERVICE_SET_COVER_POSITION,
        service_data={ATTR_ENTITY_ID: test_entity, ATTR_POSITION: 22},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closing"

    for _ in range(5):
        future = dt_util.utcnow() + datetime.timedelta(seconds=1)
        async_fire_time_changed(hass, future)
        await hass.async_block_till_done()

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "open"
    assert state.attributes == {
        "friendly_name": "Left Shade",
        "device_class": "blind",
        "current_position": 22,
        "supported_features": 7,
    }


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/garage-door-example.yaml", "cover.garage_door")],
)
async def test_garage_door(
    hass: HomeAssistant, setup_integration: None, test_entity: str
) -> None:
    """Test a garage door device."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closed"
    assert state.attributes == {
        "friendly_name": "Garage Door",
        "device_class": "garage",
        "current_position": 0,
        "supported_features": 3,
    }

    await hass.services.async_call(
        COVER_DOMAIN,
        SERVICE_OPEN_COVER,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "opening"

    for _ in range(11):
        future = dt_util.utcnow() + datetime.timedelta(seconds=1)
        async_fire_time_changed(hass, future)
        await hass.async_block_till_done()

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "open"
    assert state.attributes == {
        "friendly_name": "Garage Door",
        "device_class": "garage",
        "current_position": 100,
        "supported_features": 3,
    }

    await hass.services.async_call(
        COVER_DOMAIN,
        SERVICE_CLOSE_COVER,
        service_data={ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    await hass.async_block_till_done()
    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closing"

    for _ in range(11):
        future = dt_util.utcnow() + datetime.timedelta(seconds=1)
        async_fire_time_changed(hass, future)
        await hass.async_block_till_done()

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closed"
    assert state.attributes == {
        "friendly_name": "Garage Door",
        "device_class": "garage",
        "current_position": 0,
        "supported_features": 3,
    }


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/gate-example.yaml", "cover.driveway_gate")],
)
async def test_gate(
    hass: HomeAssistant, setup_integration: None, test_entity: str
) -> None:
    """Test a gate device as a cover entity."""

    state = hass.states.get(test_entity)
    assert state
    assert state.state == "closed"
    assert state.attributes == {
        "friendly_name": "Driveway Gate",
        "device_class": "gate",
        "current_position": 0,
        "supported_features": 3,
    }


@pytest.mark.parametrize(
    ("config_yaml_fixture", "test_entity"),
    [(f"{FIXTURES}/smart-blinds-example.yaml", "cover.left_shade")],
)
@pytest.mark.parametrize(
    ("device_state"),
    [
        (device_state)
        for device_state in ("open", "closed")
    ],
)
async def test_hvac_device_state(
    hass: HomeAssistant,
    setup_integration: None,
    test_entity: str,
    device_state: str,
    config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Test an HVAC device with restorable state."""

    await restore_state(hass, config_entry, "Family Room", "Left Shade", device_state)
    state = hass.states.get(test_entity)
    assert state
    assert (state.state, state.attributes) == snapshot
