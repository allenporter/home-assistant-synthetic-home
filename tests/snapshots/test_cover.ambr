# serializer version: 1
# name: test_hvac_device_state[tests/fixtures/smart-blinds-closed.yaml]
  tuple(
    'closed',
    ReadOnlyDict({
      'current_position': 0,
      'device_class': 'blind',
      'friendly_name': 'Left Shade',
      'supported_features': <CoverEntityFeature.OPEN|CLOSE|SET_POSITION: 7>,
    }),
  )
# ---
# name: test_hvac_device_state[tests/fixtures/smart-blinds-open.yaml]
  tuple(
    'open',
    ReadOnlyDict({
      'current_position': 100,
      'device_class': 'blind',
      'friendly_name': 'Left Shade',
      'supported_features': <CoverEntityFeature.OPEN|CLOSE|SET_POSITION: 7>,
    }),
  )
# ---
